### PR TITLE
7.2 Developer | UAD Erasure UI Enhanacements| LRDOCS-7012

### PR DESCRIPTION
**Pull request by rbohl**
https://issues.liferay.com/browse/LRDOCS-7012, thanks for the draft, Drew. Let me know how this looks. In particular, please review my code examples for the article on implementing search and filtering. I wholesale stole them from the test class DummyUADDisplay.

### CODE REVIEW

File: developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+- `keywords`---this will be whatever is typed into the search bar. The search
+    method should filter by whatever fields are relevant for the given entity.
+
+- `orderByField`---the name of the field used to sort the results. This will be
+    one of the names returned by `getSortingFieldNames`.
+
+- `orderByType`---whether to sort the results in ascending order or descending
+    order.  Will be one of "asc" or "desc". For pagination.
+
+- `end`---the ending index of the result set. For pagination.
```
Comment (drewbrokke) [inline developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown:62]: Sorry if I missed this earlier, but there should be a `start` parameter before `end`. It should read as follows:
```markdown
- `start` ---the starting index of the result set. For pagination.
```

File: developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+`getParentContainerClass` returns the class of the type that contains this type.
+It can return itself (for example, a folder can contain a folder).
+
+`getParentContainerId` returns the primary key of the container that contains
+the entity passed to this method.
+
+`isUserOwned` returns whether or not the given entity is owned by the user
+
+Additionally, implement this method in the `*UADDisplay` class of all containers:
```
Comment (drewbrokke) [inline developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown]: I would reword this to say:
```markdown
Additionally, implement this method in the `*UADDisplay` class for all types classified as _containers_:
```

File: developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+`getTopLevelContainer` is used to derive the count of how many user-owned
+entities are contained inside a given container's tree.
+
+It answers the question "which type `T` ancestor of `childObject` is an
+immediate child of the container identified by `parentContainerClass` and
+`parentContainerId`?" The method may return null, if `childObject` is not a
+child of the parent container at all. It's method is the most complicated to
```
Comment (drewbrokke) [inline developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown]: `It's` -> `This`

File: developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+implement and requires some consideration for each case. Refer to the test case
+for examples of the requirements used for `DLFolder`:
+[DLFolderUADDisplayTest#testGetTopLevelContainer](https://github.com/liferay/liferay-portal/blob/c8f78609353d6a83a0b755b0bbf93764959821ee/modules/apps/document-library/document-library-uad-test/src/testIntegration/java/com/liferay/document/library/uad/display/test/DLFolderUADDisplayTest.java#L67)
+
+See the actual implementation for `DLFolder` in [DLFolderUADDisplay#getTopLevelContainer](https://github.com/liferay/liferay-portal/blob/c8f78609353d6a83a0b755b0bbf93764959821ee/modules/apps/document-library/document-library-uad/src/main/java/com/liferay/document/library/uad/display/DLFolderUADDisplay.java#L105).
+
+The method returns either `null` or the `childObject` (could be any `Object`)
+representing the top level container that holds entities for the given User in
+this particular UAD task, so that the User's entities inside the container can
+be counted.
```
Comment (drewbrokke) [inline developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown]: I don't think this paragraph is accurate. May have been my mistake.

```markdown
The method returns either `null` or the container object of type T that is the top level container of the `childObject` (which could be any type of object that is a part of the hierarchy). This container does not necessarily have to be owned by the user, but is understood to contain data related to the user. This information is used to count how much user data is contained inside the container designated by `parentContainerClass` and `parentContainerId`.
```

Let me know if that seems ok to you, @rbohl

File: developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+public boolean isSiteScoped() {
+    Class<FooEntry> typeClass = FooEntry.class;
+
+    if (GroupedModel.class.isAssignableFrom(typeClass)) {
+        return true;
+    }
+
+    return false;
+}
+```
```
Comment (drewbrokke) [inline developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown:31]: I personally wouldn't recommend this particular code example since it's designed to be a catch-all for most cases, but also is designed to be overridden.  A better example would be to simply return `true` or `false` (which is probably going to be the most common case in a custom implementation).

File: developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+## Search
+
+Implement the `search` and `searchCount` methods to enable search in the UAD
+interface:
+
+<!--Is this realistic dummy code for third party devs? -->
+<!-- Straight out of DummyUADDisplay.java -->
```
Comment (drewbrokke) [inline developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: It is ok, but we should encourage the devs to actually implement some sort of real search (with filtering) if possible.  The `DummyUADDisplay` was designed to test other capabilities, not to template what a search implementation should actually look like.

**Review by drewbrokke — CHANGES_REQUESTED** [review]
Hey Russ, overall this looks good.  There are a few things that need changing, but otherwise the draft looks accurate to me, with the only exception being the explanation of the return value for `getTopLevelContainer` (which, admittedly, is _extremely_ easy to get wrong).

File: developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown (hunk 4 above)
Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown]: Nah, I think I tried to read the code myself here, so it's my mistake. :grin: Replaced with your explanation.

File: developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown (hunk 5 above)
Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown:31]: Okay, I'll update the method description to read

> The `isSiteScoped` method returns a boolean based on whether or not the entities can be associated with a particular site: return `false` if not, and `true` if the entities are scoped to a site. This is used to determine which filter they will be associated with ("instance", "personal-site", or "regular-sites").

And the code example to just be 

```java
@Override
public boolean isSiteScoped() {

    return false;
}
```

:+1:  or :-1:  ? cc @drewbrokke

File: developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown (hunk 6 above)
Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: Do you mean, code to call the entity's indexer.search?

Create a searchContext (`SearchContextFactory.getInstance`), get the Indexer (`IndexerRegistryUtil.getIndexer(Entry.class)`, then execute the indexer.search?

There's no example of this in any portal app's UADDisplay that I can see, so I could use some example code if possible. Especially to create the search context, which usually requires a request, I believe?
@drewbrokke

**Comment by rbohl** [comment]
@drewbrokke  thanks for the review! Just a couple more questions to get ironed out and this will be ready to go, I think.

File: developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown (hunk 6 above)
Comment (drewbrokke) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: Yes, that kind of example would work well.  We went for trying to cover as many cases as possible inside of the core product, and if a customer follows our ServiceBuilder-based instructions they'll end up with that base case code OOTB. But if they want to implement this method themselves, it's likely the'll want to use the indexer if their entity has it.  Ultimately they can use any approach they want so long as the `search` method returns items that are related to the user and match the given search parameters. What that means exactly is up to each app.

Also, `SearchContext` doesn't require a request, you may be thinking of `SearchContext.getInstance`.  However, there's a different `SearchContext.getInstance` that doesn't require a request but instead asks for individual parameters (see https://github.com/liferay/liferay-portal/blob/22502f6df640d5b33f4d507354e327174f34159e/portal-kernel/src/com/liferay/portal/kernel/search/SearchContextFactory.java#L120).

Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: @drewbrokke 
How does this look for example code? I don't know where the Layout or the groupId would come from in this context, maybe you can help me with that:

```java
@Override
public List<T> search(
    long userId, long[] groupIds, String keywords, String orderByField,
    String orderByType, int start, int end) {

    SearchContext searchContext = SearchContextFactory.getInstance(
        new long[0], new String[0], null, PortalUtil.getDefaultCompanyId(),
        keywords, NEED LAYOUT,
        LocaleThreadLocal.getSiteDefaultLocale(), NEED GROUPID,
        TimeZoneThreadLocal.getDefaultTimeZone(), userId);

    searchContext.setStart(start);
    searchContext.setEnd(end);

    Indexer indexer = IndexerRegistryUtil.getIndexer(FooEntry.class);

    Hits hits = indexer.search(searchContext);

    List<FooEntry> fooEntries = new ArrayList<FooEntry>();

    for (int i = 0; i < hits.getDocs().length; i++) {
            Document doc = hits.doc(i);
            long entryId = GetterUtil
            .getLong(doc.get(Field.ENTRY_CLASS_PK));
            Entry entry = null;

            try {
                    entry = _fooEntryLocalService.getFooEntry(fooEntryId);
            } catch (PortalException pe) {
                    _log.error(pe.getLocalizedMessage());
            } catch (SystemException se) {
                    _log.error(se.getLocalizedMessage());
            }

            fooEntries.add(fooEntry);
    }
    return fooEntries;
}
```

Comment (drewbrokke) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: @rbohl unfortunately we're at the edge of my authority, from here on out you get my best guesses ;) 

I don't know about `Layout` in this context, but for `groupId` you can just use 0, and then use `searchContext.setGroupIds(groupIds)` after the searchContext is created.

You'd ultimately be better off talking to one of the members of the search team to come up with a solid example.

Comment (drewbrokke) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: @rbohl on second look, I don't see any usages of `searchContext.getLayout`, so as far as sample code is concerned you're pretty safe to pass in pretty much anything for the `Layout` at this point.

Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: Thanks Drew.

@lipusz do you have any input about the example`indexer.search` call I've cobbled together in the above code snippet. I'll take Drew's suggestion about the groupIds, but haven't added it yet. The context is, we're wanting to show sample code for writing the `search` method of the UADDisplay interface.

Comment (lipusz) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: @rbohl 

You can just use 
```java
SearchContext searchContext = new SearchContext();
```

and then set what you need. You don't need a `Layout` here.

Comment (lipusz) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: One question regarding the API `UADDisplay.search(long, long[], String, String, String, int, int)`:
> `@param  userId the primary key of the user`

So `userId` refers to the `User` the invoker wants to retrieve the data associated with or it's the `userId` of the invoker User?

Comment (lipusz) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: I'm asking because `searchContext.setUser(long userId)` is used to include the permission "pre-filters" in the query to filter the results by VIEW permission so here you'd need to `userId` of the invoker `User`.

Otherwise you need to post-filter the results.

Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: Good question. I think this is the User ID not of the invoker, but of the User whose data is being sanitized. @drewbrokke ?

Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: @lipusz and @drewbrokke 

Here's what I have right now, for populating the searchContext:

```java
    SearchContext searchContext = new SearchContext();

    searchContext.setStart(start);
    searchContext.setEnd(end);
    searchContext.setGroupIds(groupIds);
    searchContext.setKeywords(keywords);
    searchContext.setUserId(???);

    Indexer indexer = IndexerRegistryUtil.getIndexer(FooEntry.class);

    Hits hits = indexer.search(searchContext);
```

As you can see, not sure what to pass as the userId, since the userId we have here is the ID of the user whose data we're searching for. If we don't pass anything, Tibor mentioned we'll have to do post-filtering. I'm a bit stuck as to how I proceed here. We're in the Control Panel, using admin only features. Not sure if that matters.
:thinking:

Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: Does this get me anywhere? Idea stolen from the powwow portlet's ZoomAPICallbackServlet class :man_shrugging: 

```java
    BooleanQuery booleanQuery = BooleanQueryFactoryUtil.create(
        searchContext);

    booleanQuery.addExactTerm("userId", userId);

    BooleanClause booleanClause = BooleanClauseFactoryUtil.create(
        booleanQuery, BooleanClauseOccur.MUST.getName());

    searchContext.setBooleanClauses(new BooleanClause[] {booleanClause});
```

Comment (rbohl) [reply developer/frameworks/articles/user-associated-data/04-filtering-and-searching.markdown]: @drewbrokke I'm gonna take this directly to the search team via the #help-search channel in slack (or #t-search; so many choices). That way we don't just overload Tibor, but give the whole team a chance to respond as they have time.

**Comment by rbohl** [comment]
@drewbrokke can you review this commit to make sure it checks out? https://github.com/drewbrokke/liferay-docs/pull/8/commits/1baebbf890e52bdd914d4264454ea332dc822680

**Comment by rbohl** [comment]
> Perhaps we should also point out the significance of `groupIds` and `keywords` as filtering parameters. Are we relying on the Javadocs to point this out, or should it also be part of the tutorial?

@drewbrokke  We say this in the more explanatory article: 

>- `groupIds`---this is an array of groupIds used to filter which data is shown
>    by which groups it is associated with. In the case that no groupIds are
>    given (it can be null), the search method should return data that is not
>    scoped to any given group.
>
> - `keywords`---this will be whatever is typed into the search bar. The search
>    method should filter by whatever fields are relevant for the given entity.

Good enough?

**Comment by rbohl** [comment]
Superceded: https://github.com/sez11a/liferay-docs/pull/4442
